main.go

```diff
@@ -91,26 +102,56 @@ func main() {
 			tlsConn, err := vhost.TLS(c)
 			if err != nil {
 				log.Printf("Error accepting new connection - %v", err)
+				return
 			}
+			defer tlsConn.Close()
+
 			if tlsConn.Host() == "" {
 				log.Printf("Cannot support non-SNI enabled clients")
 				return
 			}
-			connectReq := &http.Request{
-				Method: "CONNECT",
-				URL: &url.URL{
-					Opaque: tlsConn.Host(),
-					Host:   net.JoinHostPort(tlsConn.Host(), "443"),
-				},
-				Host:   tlsConn.Host(),
-				Header: make(http.Header),
+
+			remoteHost := net.JoinHostPort(tlsConn.Host(), "443")
+			conn, err := dialer.Dial("tcp", remoteHost)
+			if err != nil {
+				log.Printf("Error establishing connection to %s: %v", remoteHost, err)
+				return
 			}
-			resp := dumbResponseWriter{tlsConn}
-			proxyServer.ServeHTTP(resp, connectReq)
+			defer conn.Close()
+
+			if proxyConnection, ok := conn.(*proxy.Connection); ok {
+				clientHost, _, err := net.SplitHostPort(c.RemoteAddr().String())
+				if err != nil {
+					log.Printf("Failed to get host from address %s: %v", c.RemoteAddr(), err)
+					return
+				}
+				if err := proxyConnection.ConnectTo(conn, remoteHost, sm.Hash(clientHost)); err != nil {
+					log.Printf("Error establishing CONNECT tunnel to %s: %v", remoteHost, err)
+					return
+				}
+			}
```

Comment:
If we create our custom `Dialer` interface this condition just will be in another place, but we still need it since we do different dials for proxy traffic, and for traffic to originaly requested server. 
And `goproxy` requires standard `Dialer` interface. We cannot replace it now. 
Anyway, this sounds more like a future improvement.